clippy_lints/src/assertions_on_constants.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2014-2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::rustc::hir::{Expr, ExprKind};
+use crate::rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use crate::rustc::{declare_tool_lint, lint_array};
+use crate::syntax::ast::LitKind;
+use crate::utils::{is_direct_expn_of, span_lint, span_lint_and_sugg};
+use rustc_errors::Applicability;
+use if_chain::if_chain;
+
+/// **What it does:** Check explicit call assert!(true/false)
+///
+/// **Why is this bad?** Will be optimized out by the compiler or should probably be replaced by a panic!() or unreachable!()
+///
+/// **Known problems:** None
+///
+/// **Example:**
+/// ```rust
+/// assert!(false)
+/// // or
+/// assert!(true)
+/// // or
+/// const B: bool = false;
+/// assert!(B)
+/// ```
+declare_clippy_lint! {
+    pub ASSERTIONS_ON_CONSTANTS,
+    style,
+    "assert!(true/false) will be optimized out by the compiler/should probably be replaced by a panic!() or unreachable!()"
+}
+
+pub struct AssertionsOnConstants;
+
+impl LintPass for AssertionsOnConstants {
+    fn get_lints(&self) -> LintArray {
+        lint_array![ASSERTIONS_ON_CONSTANTS]
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for AssertionsOnConstants {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        if_chain! {
+            if is_direct_expn_of(e.span, "assert").is_some();
+            if let ExprKind::Unary(_, ref lit) = e.node;
+            if let ExprKind::Lit(ref inner) = lit.node;
+            then {
+                match inner.node {
+                    LitKind::Bool(true) => {
+                        span_lint(cx, ASSERTIONS_ON_CONSTANTS, e.span,
+                            "assert!(true) will be optimized out by the compiler");
+                    },
+                    LitKind::Bool(false) => {
+                        span_lint_and_sugg(
+                            cx,
+                            ASSERTIONS_ON_CONSTANTS,
+                            e.span,
+                            "assert!(false) should probably be replaced",
+                            "try",
+                            "panic!()".to_string(),
```

Comment:
`panic!()` or `unreachable!()`

Reply:
Work in progress. I pushed the current state of PR to test span_lint_and_sugg. Seems it not work as expected for `assert!()` macro. Maybe I should replace it with `span_lint_and_help` function?